gdk4/src/event.rs

```diff
@@ -0,0 +1,616 @@
+            }
+        }
+    }
+
+    #[doc(alias = "gdk_event_get_seat")]
+    pub fn get_seat(&self) -> Option<Seat> {
+        unsafe { from_glib_none(ffi::gdk_event_get_seat(self.to_glib_none().0)) }
+    }
+
+    #[doc(alias = "gdk_event_get_surface")]
+    pub fn get_surface(&self) -> Option<Surface> {
+        unsafe { from_glib_none(ffi::gdk_event_get_surface(self.to_glib_none().0)) }
+    }
+
+    #[doc(alias = "gdk_event_get_time")]
+    pub fn get_time(&self) -> u32 {
+        unsafe { ffi::gdk_event_get_time(self.to_glib_none().0) }
+    }
+
+    #[doc(alias = "gdk_event_triggers_context_menu")]
+    pub fn triggers_context_menu(&self) -> bool {
+        unsafe { from_glib(ffi::gdk_event_triggers_context_menu(self.to_glib_none().0)) }
+    }
+}
+
+impl fmt::Display for Event {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("Event")
+    }
+}
+
+pub trait EventTypeTrait: StaticType + FromGlibPtrFull<*mut ffi::GdkEvent> + 'static {
+    fn event_types() -> Vec<EventType>;
+}
+
+macro_rules! define_event {
+    ($rust_type:ident, $ffi_type:path, $ffi_type_path:path, $event_event_types:expr) => {
+        // Can't use get_type here as this is not a boxed type but another fundamental type
+        glib::wrapper! {
+            #[derive(Debug)]
+            pub struct $rust_type(Shared<$ffi_type>);
+
+            match fn {
+
+                ref => |ptr| ffi::gdk_event_ref(ptr as *mut ffi::GdkEvent) as *mut $ffi_type,
+                unref => |ptr| ffi::gdk_event_unref(ptr as *mut ffi::GdkEvent),
+            }
+        }
+
+        impl StaticType for $rust_type {
+            fn static_type() -> Type {
+                unsafe { from_glib($ffi_type_path()) }
+            }
+        }
+
+        impl EventTypeTrait for $rust_type {
+            fn event_types() -> Vec<EventType> {
+                $event_event_types
+            }
+        }
+
+        impl std::ops::Deref for $rust_type {
+            type Target = Event;
+
+            fn deref(&self) -> &Self::Target {
+                unsafe { &*(self as *const $rust_type as *const Event) }
+            }
+        }
+
+        impl AsRef<Event> for $rust_type {
+            fn as_ref(&self) -> &Event {
+                self.upcast_ref()
+            }
+        }
+
+        #[doc(hidden)]
+        impl FromGlibPtrFull<*mut ffi::GdkEvent> for $rust_type {
+            unsafe fn from_glib_full(ptr: *mut ffi::GdkEvent) -> Self {
+                from_glib_full(ptr as *mut $ffi_type)
+            }
+        }
+
+        impl $rust_type {
+            pub fn upcast(self) -> Event {
+                unsafe { from_glib_full(self.to_glib_full() as *mut ffi::GdkEvent) }
+            }
+
+            pub fn upcast_ref(&self) -> &Event {
+                &*self
+            }
+        }
+
+        impl fmt::Display for $rust_type {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                f.write_str(stringify!($rust_type))
```

Comment:
Even nicer if these would print all the data of the event, same for the `Debug` impl